migrate to docusaurus (v2)

Woile opened the pull request:
Based on the work from @ibraheemdev  I fixed the conflicts, still have to take a look at the comments from that PR and fix the jumbotron, atm it looks like this:

<img width="1626" alt="Screenshot 2022-03-18 at 11 44 43" src="https://user-images.githubusercontent.com/2585816/158989116-3cb949eb-5b5f-4224-a437-33867d3568a5.png">

Continuation of #222 

## Evaluation

1. Checkout to this PR
2. Run `npm install && npm start`
3. Navigate the docs

## Demo

https://woile.github.io/actix-website/

robjtede commented:
Doesn't have to be the same cover image (jumbotron) if something else works better for the new style. Just make sure whatever you go with is permissively licensed.

Woile commented:
I think I've covered all the suggestions in the original PR, let me know of more.
I've skipped LICENSE, I'll leave it to someone more experienced with actix licencing.

TODO:

- [ ] Update github action
- [ ] Add github link at the upper right corner to actix-web (as a user of the website I'm constantly trying to find the link to github), but we leave the "Code" section for the other repos.
- [ ] Should API documentation include links to experiments and extras? I think it would be nice to include at least extras there
- [ ] Should we keep actix link in API documentation? asking out of curiosity, as I don't usually use it

nikstur commented on LICENSE-MIT:
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2022 Actix Team
+Copyright (c) 2017-2021 Actix Team
```
I would keep the license changes out of this PR. See also LICENSE-APACHE

Woile replied:
Removed from the original commits from @ibraheemdev the changes to the LICENCES

nikstur commented on README.md:
```diff
@@ -2,43 +2,39 @@
 
-[hugo]: https://gohugo.io
-[homebrew]: https://brew.sh
-[mermaid cli]: https://github.com/mermaidjs/mermaid.cli
```
Removing this breaks the Mermaid CLI link in line 18.

nikstur commented on examples/application/Cargo.toml:
```diff
@@ -5,4 +5,5 @@ edition = "2018"
 
 [dependencies]
-actix-web = "4"
+actix-web = "3"
```
Looks like you might have copied some old example code from the other that has been updated in the meanwhile to version 4 of actix-web. Maybe it's best to not change the examples at all in this PR.

nikstur commented on examples/Cargo.toml:
```diff
@@ -23,5 +23,4 @@ members = [
 ]
-# TODO: update databases
-exclude = ["databases"]
+exclude = ["databases", "sentry"]
```
How does this apply here? There is no sentry example crate, right?

nikstur commented on .gitignore:
```diff
@@ -1,8 +1,4 @@
-public
-tmp
-_site
-Cargo.lock
```
I think we should continue to keep Cargo.lock out of git. Thus, keep it in the .gitignore ;)

nikstur reviewed (changes requested):
Looks great! 
Only issue that block this are the changes to the examples. Many of the changes made for version 4, were reverted by you (by copying the changes from the other PR, I assume).

- [ ] Fix examples

nikstur commented:
About your TODOs: I think only the Github Actions changes are necessary for this PR. The other TODOs might very well be worth doing, but they can be done in separate PRs after the initial migration to docusaurus is done.

Woile commented:
mmm I didn't do the rebase properly it seems then.
It makes sense to postpone the todos till after merging the PR.

I'm gonna do a pass over the current docs and fix the issues

Woile commented:
I modified the original commit "migrate to docusaurus" and removed the changes to the examples folder (while keeping the author), I think the results are good 💪🏻 (I'm still going through all of the sections comparing with the current website)

Woile commented:
the websockets one is still failing, not sure why, I think is also failing in master

Woile commented:
Some extra stuff to do (can be in next PR):

- [ ] fix code in the home, I see that in the hugo version it is hardcoded, at the moment is reading from the packages (I aligned the packages to master, so at least now it looks similar to the hugo version)
- [ ] make code readable in the home (making the code box bigger, it doesn't make sense to have overflow there)
- [ ] insert the code snippet of the main-example in the home

nikstur commented on examples/Cargo.lock:
```diff
@@ -0,0 +1,3088 @@
+# This file is automatically @generated by Cargo.
```
This file has probably slipped in before you readded Cargo.lock to .gitignore.

Woile commented:
I've changed the colors to give a bit more personality to the website, do let me know if it sucks 😅

nikstur commented:
Looks good to me now. I opened  #267 to fix the failing websocket example. If we get that merged, your checks shouldn't fail anymore and we can merge this PR too.

robjtede commented:
Is this ready for a proper review now?

Woile commented:
Yes! Go ahead! 💪🏻

Woile commented:
Is it still planned to merge this? should I rebase?

robjtede commented:
Yeah this migration is still on the cards.

Woile commented:
Rebased!

JohnTitor commented:
Since this is a major migration, I'd like to have some periods during migration and make sure everything is fine.
We could:
- host this on a subdomain and show a message like "we're moving to a new website, https://new.actix.rs" on actix.rs, then do an actual migration and redirection from the subdomain after some periods.
- host the current on prev.actix.rs or something else.

JohnTitor commented on .github/workflows/test-deploy.yml:
```diff
@@ -0,0 +1,24 @@
+      - name: Install dependencies
+        run: npm install
+      - name: Test build website
+        run: npm run build
```
This workflow doesn't seem to deploy anything but just builds it. I guess we could use Vercel or something else for a preview deployment.

Woile replied:
Indeed, using vercel/netlify would be great. But even though it does nothing (deployment wise), it can still catch errors, specially from docusarus.

JohnTitor replied:
Then it should be renamed to build-check or something else, the current name is kinda confusing.

JohnTitor commented on vars.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  rustVersion: '1.54',
```
Now it's 1.56 (or 1.57, it's under review).

JohnTitor commented on examples/websockets/src/main.rs:
```diff
@@ -41,4 +41,4 @@ async fn main() -> std::io::Result<()> {
 // Connection: Upgrade
 // Sec-WebSocket-Key: SOME_KEY
-// Sec-WebSocket-Version: 13
+// Sec-WebSocket-Version: 13
```
In general we prefer to break a new line at the end of the file.

JohnTitor commented:
> Should API documentation include links to experiments and extras? I think it would be nice to include at least extras there

For extras, yes I'd prefer. For experiments, I'd like to ask @robjtede.

> Should we keep actix link in API documentation? asking out of curiosity, as I don't usually use it

Considering the domain is actix.rs and it's our official crate, I'd like to have it on the website.

Woile commented:
I'm gonna try to deploy to my github pages and share the link if I manage to do it.

Having an integration with netlify/vercel would be great for previews (out of my control).

JohnTitor commented:
Thanks! Skimmed it and almost LGTM. I've found two issues that I'd like to fix on this PR.

First, the "Edit this page" link doesn't work, e.g. it provides a link https://github.com/actix/actix-website/docs/whatis.md but I think it should be https://github.com/actix/actix-website/edit/master/docs/whatis.md.

And there's a style issue on https://woile.github.io/actix-website/docs/url-dispatch/:

![image](https://user-images.githubusercontent.com/25030997/174791753-45f5dd96-4e3c-45ed-bbe9-1acb9973b6ce.png)

[The current](https://actix.rs/docs/url-dispatch/) looks like:

![image](https://user-images.githubusercontent.com/25030997/174791980-f394937c-01c8-417a-a14a-8ec5d72ddf34.png)

notice `/`'s styling is weird. I checked on Firefox and Chrome and I think it's a valid issue. Is that fixable on our side somehow?

JohnTitor commented:
> Let me know if it's better now

Confirmed the above two issues have been fixed, thanks :)

JohnTitor commented:
@robjtede Any thoughts on https://github.com/actix/actix-website/pull/266#issuecomment-1161491197?
I think it's a robust way but on the other hand, I slightly feel it may be too cautious.

robjtede commented:
Maybe. We'd have to email Nikolay for adding subdomains though.

Woile commented:
> Linking to experiments: sure but probably best to link to the docs.rs pages to ensure link permanence.

I will skip new features for now. We already have a lot 😅 I prefer to introduce new things later in the future.
I think we should finish the migration first.

robjtede commented:
The diagrams are broken on my local build?

Woile commented:
In the hosted version seems to be working, could you show me how it looks in your local or if any errors appear?

robjtede commented:
```
 › npm start

> actix-website@2.0.0 start
> docusaurus start

[INFO] Starting the development server...
[SUCCESS] Docusaurus website is running at: http://localhost:3000/actix-website/

✖ Client
  Compiled with some errors in 1.43s



Error: error:0308010C:digital envelope routines::unsupported
    at String.replace (<anonymous>)
Error: error:0308010C:digital envelope routines::unsupported
    at String.replace (<anonymous>)
Error: error:0308010C:digital envelope routines::unsupported
    at String.replace (<anonymous>)
Error: error:0308010C:digital envelope routines::unsupported
    at String.replace (<anonymous>)
Error: error:0308010C:digital envelope routines::unsupported
    at String.replace (<anonymous>)
client (webpack 5.72.1) compiled with 5 errors
```

And the browser overlays errors as such:
```
Compiled with problems:

ERROR in ./static/img/diagrams/connection_accept.svg (./node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/Users/rob/Development/forks/actix-website/node_modules/file-loader/dist/cjs.js!./static/img/diagrams/connection_accept.svg)

Module build failed (from ./node_modules/url-loader/dist/cjs.js):
Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:67:19)
    at Object.createHash (node:crypto:133:10)
    at getHashDigest (/Users/rob/Development/forks/actix-website/node_modules/loader-utils/lib/getHashDigest.js:46:34)
    at /Users/rob/Development/forks/actix-website/node_modules/loader-utils/lib/interpolateName.js:113:11
    at String.replace (<anonymous>)
    at interpolateName (/Users/rob/Development/forks/actix-website/node_modules/loader-utils/lib/interpolateName.js:110:8)
    at Object.loader (/Users/rob/Development/forks/actix-website/node_modules/file-loader/dist/index.js:29:48)
    at Object.loader (/Users/rob/Development/forks/actix-website/node_modules/url-loader/dist/index.js:127:19)
```

Woile commented:
You can try one more time, not sure what happened, I added `sass` dep which was missing. I'm using the TLS version of node: `v16.14.0`

If it doesn't work I'll containerize it

robjtede commented:
node 16 works for me, no worries

JohnTitor commented:
I've created v2 branch, could you change the base branch to there? Also, I'd like to deploy the new website on Netlify so we don't need a deploy action.

Woile commented:
Updated!

JohnTitor reviewed (approved):
Thanks!

Let's merge and see if the deployment on Netlify works fine. Then I'd like to deploy this to actix.rs and move the current to prev.actix.rs or somewhere else.
